### ISSUE TITLE
`-msse4.2` should produce an error when passed to inappropriate targets

### ISSUE THREAD

**Issue by sh1boot** [llvm/llvm-project#70143]
Clang seems to accept (with a warning) `-msse4` for any target, not just x86.

No problem, as expected:

    clang-18 --target=amd64 -msse2 -c -xc /dev/null
    clang-18 --target=amd64 -msse3 -c -xc /dev/null
    clang-18 --target=amd64 -msse4 -c -xc /dev/null
    clang-18 --target=amd64 -msse4.2 -c -xc /dev/null

Error, as expected:

    $ clang-18 --target=arm64 -msse2 -c -xc /dev/null
    clang-18: error: unsupported option '-msse2' for target 'arm64'
    $ clang-18 --target=arm64 -msse3 -c -xc /dev/null
    clang-18: error: unsupported option '-msse3' for target 'arm64'

Warning, this is unexpected behaviour:

    $ clang-18 --target=arm64 -msse4 -c -xc /dev/null
    clang-18: warning: argument unused during compilation: '-msse4' [-Wunused-command-line-argument]
    $ clang-18 --target=arm64 -msse4.2 -c -xc /dev/null
    clang-18: warning: argument unused during compilation: '-msse4.2' [-Wunused-command-line-argument]

This breaks feature detection in some cmake files which do not see a failure from clang, meaning sse4.2 appears to be available for the compilation target.

(unexpected behaviour exists for arm64 and riscv64 at least; I haven't tested others)

**Comment by topperc** [comment]
This may be related to this FIXME

```
// TODO: Make -msse4.2 TargetSpecific after                                      
// https://github.com/llvm/llvm-project/issues/63270 is fixed.                   
def msse4_2 : Flag<["-"], "msse4.2">, Group<m_x86_Features_Group>;
```

@MaskRay

**Comment by MaskRay** [comment]
Thanks for using this feature:) 

`-msse4.2` remains a warning as a workaround for `compiler-rt/lib/tsan` on macOS. I hopefully someone with macOS can investigate https://github.com/llvm/llvm-project/issues/63270

**Comment by sh1boot** [comment]
I contemplated reporting this as a cmake bug, too, because what broke for me was that [`check_c_compiler_flag()`](https://cmake.org/cmake/help/latest/module/CheckCCompilerFlag.html) approved the flag despite clang issuing a [warning-level] diagnostic.  I can't decide if this is in agreement with cmake documentation or not, but it does cause really noisy builds (as well as broken builds).

However, if that behaviour changed I think I would also break the workaround mentioned above, so I guess in the interim I should not mention it.  🤷🏽

**Comment by sh1boot** [comment]
As a workaround, calling `cmake` with `CFLAGS=" [...] -Werror=unused-command-line-argument"` will stop it trying to enable nonsense code.

Hopefully that shouldn't conflict with whatever else is going on in the build system.  I know it'll conflict when you put things like `-fuse-ld=xxx` in `CFLAGS` instead of `LDFLAGS`, but I guess just try to not get into that situation in the interim.

**Comment by MaskRay** [comment]
> As a workaround for people whose cmake configurations are being poisoned by this bug; calling `cmake` with `CFLAGS=" [...] -Werror=unused-command-line-argument"` will stop it trying to enable nonsense code.
> 
> Hopefully that shouldn't conflict with whatever else is going on in the build system. I know it'll conflict when you put things like `-fuse-ld=xxx` in `CFLAGS` instead of `LDFLAGS`, but I guess just try to not get into that situation in the interim.

If you can work around #63270 (`compiler-rt/lib/tsan/rtl/CMakeLists.txt` for macOS), we can drop the workaround in clang driver...
